zetaclient/cctx_scanner.go

```diff
@@ -0,0 +1,229 @@
+// Re-check and update missed cctx's status
+func (sc *CctxScanner) UpdateMissedPendingCctxStatus(chainID int64, nonce uint64) {
+	send, err := sc.bridge.GetCctxByNonce(chainID, nonce)
+	if err != nil {
+		sc.logger.Error().Err(err).Msgf("scanner: error GetCctxByNonce for chain %d nonce %d", chainID, nonce)
+		return
+	}
+	// A missed cctx will pend forever if:
+	//    1. No tracker.   For some reason(e.g., RPC failure), no observer had reported outtx hash to zetacore.
+	//    2. No true hash. Track exists but none of the hashes is true (can't be verified)
+	if !crosschaintypes.IsCctxStatusPending(send.CctxStatus.Status) { // no longer pending
+		sc.removeMissedPendingCctx(chainID, nonce)
+		sc.logger.Info().Msgf("scanner: removed missed pending cctx for chain %d nonce %d", chainID, nonce)
+	}
+}
+
+func (sc *CctxScanner) addMissedPendingCctx(chainID int64, nonceFrom uint64, nonceTo uint64, missedList []*crosschaintypes.CrossChainTx) {
+	// initialize missed cctx map if not done yet
+	if _, found := sc.missedPendingCctx[chainID]; !found {
+		sc.missedPendingCctx[chainID] = make(map[uint64]*crosschaintypes.CrossChainTx)
+	}
+
+	nonces := make([]uint64, 0) // for logging only
+	for _, send := range missedList {
+		nonce := send.GetCurrentOutTxParam().OutboundTxTssNonce
+		nonces = append(nonces, nonce)
+		sc.missedPendingCctx[chainID][nonce] = send
+	}
+	sc.nextNonceToScan[chainID] = nonceTo
+	if len(missedList) > 0 {
+		sc.saveFirstNonceToScan(chainID)
+		sc.logger.Info().Msgf("scanner: found missed pending cctx for chain %d with nonces %v", chainID, nonces)
+	}
+}
+
+func (sc *CctxScanner) removeMissedPendingCctx(chainID int64, nonce uint64) {
+	delete(sc.missedPendingCctx[chainID], nonce)
+	sc.saveFirstNonceToScan(chainID)
+}
+
+func (sc *CctxScanner) saveFirstNonceToScan(chainID int64) {
+	firstNonceToScan := uint64(math.MaxUint64)
+	if len(sc.missedPendingCctx[chainID]) == 0 {
+		// either no missed pending cctx found so far OR last missed pending cctx removed
+		firstNonceToScan = sc.nextNonceToScan[chainID]
+	} else { // save the lowest nonce for future restart if there ARE missed pending cctx
+		for nonceMissed := range sc.missedPendingCctx[chainID] {
+			if nonceMissed < firstNonceToScan {
+				firstNonceToScan = nonceMissed
+			}
+		}
+	}
+	if firstNonceToScan < uint64(math.MaxUint64) {
+		if err := sc.db.Save(clienttypes.ToFirstNonceToScanSQLType(chainID, firstNonceToScan)).Error; err != nil {
+			sc.logger.Error().Err(err).Msgf("scanner: error writing firstNonceToScan for chain %d nonce %d", chainID, firstNonceToScan)
+		}
+	}
+}
+
+// LoadDB open sql database and load data into scanner
+func (sc *CctxScanner) LoadDB(dbpath string, memDB bool) error {
+	if _, err := os.Stat(dbpath); os.IsNotExist(err) {
+		err := os.MkdirAll(dbpath, os.ModePerm)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+	path := dbpath
+	if !memDB { // memDB is used for uint test only
+		path = fmt.Sprintf("%s/scanner", dbpath)
```

Comment:
```suggestion
		path = fmt.Sprintf("%s/scanner", dbpath)
		path = filepath.Join(dbpath, "scanner")
```

We can use `path/filepath` package for path operations

Reply:
Yeah. That's a more serious way. Let me try.